### ISSUE TITLE
Support for distributed custom input pipeline

### ISSUE THREAD

**Issue by niclasdn-axis** [tensorflow/tensorflow#29767]
Feature Request
------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04 and 18.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
No
- **TensorFlow installed from (source or binary)**:
Docker hub image using: FROM tensorflow/tensorflow:2.0.0b0-gpu-py3
- **TensorFlow version (use command below)**:
v1.12.1-3259-gf59745a381 2.0.0-beta0
- **Python version**:
Python 3.6.8
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
See prebuilt docker image
- **GPU model and memory**:
Nividia GeForce GTX TITAN-X (Maxwell)
- **Exact command to reproduce**:
None, since this is a feature request

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:

```bash
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"
```

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.
This is a feature request:
---------------------------------
In the guide for TensorFlow 2.0 at https://www.tensorflow.org/beta/guide/distribute_strategy
there is an example for using tf.distribute.Strategy with custom training loops. However this guide assumes that one uses tf.data.Dataset for the input pipeline.

I have the need for a very dynamic input pipeline since I do inference-based batch mining for triplet training very much like what is described in an article from Google found here:
FaceNet: A Unified Embedding for Face Recognition and Clustering
https://arxiv.org/pdf/1503.03832.pdf

I believe such capability is fundamental for having flexible customization capability while being able to distribute training. Fixed input pipeline can be very limiting.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

The above mentioned tutorial assumes you can do something like this with the input pipeline:
```python
with mirrored_strategy.scope():
  dataset = tf.data.Dataset.from_tensors(([1.], [1.])).repeat(1000).batch(
      global_batch_size)
  dist_dataset = mirrored_strategy.experimental_distribute_dataset(dataset)

with mirrored_strategy.scope():
  for inputs in dist_dataset:
    print(train_step(inputs))
```
Whereas my example of custom inference-based dataset reader looks more like this:
```python
[anchor_inputs, positive_inputs, negative_inputs] = get_dynamic_triplet_batches_pytorch(args, model, dataset_handler)
```
The dataset_handler is entirely custom and is used to provide candidate triplets. Then inference is run on them and if the loss is moderate (as described in the FaceNet paper from Google) the triplet is added to a final training batch. In the end that batch is returned to the training loop for the full backprop pass. However, the need to distribute a custom input pipeline is generic and not limited to this use-case.

The example is taken from code I wrote for PyTorch where the distribution strategy available is very flexible, so this input pipeline works just out-of-the-box (for distributing on GPUs on a single machine). I would like to see this be possible in Tensorflow as well (but here for all distribution strategies of course). This would make Tensorflow more attractive for more advanced use-cases for training. Right now I don't see how it is possible to set up a custom training input pipeline like this with the existing distribution strategies.

BR,
/Niclas

**Comment by guptapriya** [comment]
We have recently added a new API that should allow you to construct customized input and feed into distribution strategy `run` method: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/distribute/distribute_lib.py#L1318

Would something like that help in your use case?

**Comment by saikumarchalla** [comment]
@niclasdn-axis   Closing this issue as of now. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
